packages/opentelemetry-node/lib/detectors.js

```diff
@@ -0,0 +1,132 @@
+ * under the License.
+ */
+
+/**
+ * NOTE: when `Detector` is finally removed import only `DetectorSync` and
+ * get rid of the aliasing
+ * @typedef {import('@opentelemetry/resources').Detector} DetectorOrig
+ * @typedef {import('@opentelemetry/resources').DetectorSync} DetectorSyncOrig
+ * @typedef {DetectorOrig | DetectorSyncOrig} DetectorSync
+ */
+
+const {
+    alibabaCloudEcsDetector,
+} = require('@opentelemetry/resource-detector-alibaba-cloud');
+const {
+    awsBeanstalkDetector,
+    awsEcsDetector,
+    awsEksDetector,
+    awsEc2Detector,
+    awsLambdaDetector,
+} = require('@opentelemetry/resource-detector-aws');
+const {
+    azureAppServiceDetector,
+    azureFunctionsDetector,
+    azureVmDetector,
+} = require('@opentelemetry/resource-detector-azure');
+const {
+    containerDetector,
+} = require('@opentelemetry/resource-detector-container');
+const {gcpDetector} = require('@opentelemetry/resource-detector-gcp');
+const {
+    envDetectorSync,
+    hostDetectorSync,
+    osDetectorSync,
+    processDetectorSync,
+    serviceInstanceIdDetectorSync,
+    Resource,
+} = require('@opentelemetry/resources');
+const {log} = require('./logging');
+
+// @ts-ignore - compiler options do not allow lookp outside `lib` folder
+const ELASTIC_SDK_VERSION = require('../package.json').version;
+
+// Elastic's own detector to add distro related metadata
+/** @type {DetectorSync} */
+const distroDetectorSync = {
+    detect() {
+        // TODO: change to semconv resource attribs when
+        // `@opentelemetry/semantic-conventions` gets updated with the attribs used
+        // https://github.com/open-telemetry/opentelemetry-js/issues/4235
+        return new Resource({
+            'telemetry.distro.name': 'elastic',
+            'telemetry.distro.version': ELASTIC_SDK_VERSION,
+        });
+    },
+};
+
+/** @type {Record<string, DetectorSync | Array<DetectorSync>>} */
+const defaultDetectors = {
+    env: envDetectorSync,
+    process: processDetectorSync,
+    serviceinstance: serviceInstanceIdDetectorSync,
+    os: osDetectorSync,
+    host: hostDetectorSync,
+    container: containerDetector,
+    alibaba: alibabaCloudEcsDetector,
+    aws: [
+        awsBeanstalkDetector,
+        awsEc2Detector,
+        awsEcsDetector,
+        awsEksDetector,
+        awsLambdaDetector,
+    ],
+    azure: [azureAppServiceDetector, azureFunctionsDetector, azureVmDetector],
+    gcp: gcpDetector,
+};
+
+/**
+ * @param {Array<DetectorSync>} [detectors]
+ * @returns {Array<DetectorSync>}
+ */
+function resolveDetectors(detectors) {
+    if (detectors) {
+        detectors.push(distroDetectorSync);
+        return detectors;
+    }
+
+    const detectorsFromEnv =
+        process.env['OTEL_NODE_RESOURCE_DETECTORS'] || 'all';
+
+    if (detectorsFromEnv === 'none') {
+        return [];
+    }
+
+    const detectorKeys =
+        detectorsFromEnv === 'all'
+            ? Object.keys(defaultDetectors)
+            : detectorsFromEnv.split(',');
```

Comment:
The logic is *slightly* difference in auto-instr-node: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/dadf30828e7008df7b85c8bd7eaacb90770aac5c/metapackages/auto-instrumentations-node/src/utils.ts#L235-L244

E.g. `OTEL_NODE_RESOURCE_DETECTORS=host,os,none`  with a "none" in there the behaviour differs. Also if 'all' is *one* of the values in there.

Personally, I think it would be nice to `.trim()` each value so `OTEL_NODE_RESOURCE_DETECTORS=host, os` works. `auto-instr-node` does trim in its handling of `process.env.OTEL_NODE_ENABLED_INSTRUMENTATIONS`, FWIW.

Reply:
I've applied your suggestions. Thanks :)